### PR TITLE
Generic test in ci cd

### PR DESCRIPTION
**Pull request by RubelMozumder**
# In GitHub action `plugin_test.yaml`

### CODE REVIEW

**Comment by RubelMozumder** [comment]
@domna and @lukaspie can you update the last command for the plugin test in GitHub action comes in plugin_test.yaml for XPS and MPES what test do you want to run in ci/cd from pynxtools.

Then we can also think a generalised function from the plugin template for all the plugins.

File: .github/workflows/pytest.yml [hunk 1]
```diff
@@ -32,7 +32,7 @@ jobs:
           python -m pip install coverage coveralls
       - name: Install package
         run: |
-          pip install ".[dev]"
+          pip install ".[dev,apm,em,mpes,xps,stm,xrd]"
```
Comment (domna) [inline .github/workflows/pytest.yml:35]: We don't need the plugins here as we are just testing basic pynx functionalities.

File: .github/workflows/plugin_test.yaml [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+      - name: Install package
+        run: |
+          pip install ".[dev]"
+      - name: Install STM Plugin
+        run: |
+          git clone https://github.com/FAIRmat-NFDI/pynxtools-stm
+          cd pynxtools-stm
+          git checkout TestFromPynxtools
+          pip install .
+          pytest tests/test_reader.py::test_stm_reader
```
Comment (domna) [inline .github/workflows/plugin_test.yaml:42]: I would split this into multiple steps in the action. Also I would use the [github checkout action template](https://github.com/actions/checkout). See the section [here](https://github.com/actions/checkout?tab=readme-ov-file#checkout-multiple-repos-side-by-side) on how to checkout multiple repos.

File: .github/workflows/pytest.yml (hunk 1 above)
Comment (RubelMozumder) [reply .github/workflows/pytest.yml:35]: That came from the previous attempt. I will remove it.

File: .github/workflows/plugin_test.yaml (hunk 2 above)
Comment (RubelMozumder) [reply .github/workflows/plugin_test.yaml:42]: Thanks. I check it, it is more robust. I will implement it. would you just update what types of function you want to test in pynxtools. Then, with the new approve I can check the ci/cd and prepare the PR for final review.

Comment (domna) [reply .github/workflows/plugin_test.yaml]: Normally it should be fine if you just do `pytest .` on the plugins. That should gather all relevant tests and run them. That way the plugin writers have the greatest flexibility of implementing tests.

File: .github/workflows/pytest.yml (hunk 1 above)
Comment (RubelMozumder) [reply .github/workflows/pytest.yml]: removed!

**Comment by RubelMozumder** [comment]
TODO list from PR #306 
- [ ] ~~Modify function `test_has_correct_read_func(reader)` in `test_reader.py`~~
- [x] Add a generic test template on the Plugin template.
- [ ] Needs to track the semantics version of the NeXus definition and Pynxtools.
- [x] Check plugin packaging that must include the example files.

The implementation of the first point (putting as  is not needed as we are not testing all the readers-plugin in the `test_reader.py`. My plan was to bring the existing test for readers residing in pynxtools together with the test for plugin readers. 

For the second point, we have to add another matrix for `pynxtools` version in the ci/cd workflow. Then PyPI automatically grap the compatible version for pynxtools. Still, I can not try it in `sts` reader. Once this PR is accepted, I will go and test it to include the semantic version. I have created another PR for that: #319

File: pynxtools/test_suite/reader_plugin.py
Comment (domna) [inline pynxtools/test_suite/reader_plugin.py:1]: I find the naming a bit confusing. We are not testing a reader plugin we provide a generic test for reading files and converting those to nexus automatically. Also it is not really a `test_suit`. Maybe just something like `testing/nexus_conversion.py`?

File: pyproject.toml [hunk 3]
```diff
@@ -69,12 +69,6 @@ apm = [
 em = [
     "pynxtools-em",
 ]
-mpes = [
-    "pynxtools-mpes",
-]
-xps = [
-    "pynxtools-xps",
-]
```
Comment (domna) [inline pyproject.toml:77]: This should be reenabled before we actually merge this.

File: pynxtools/test_suite/reader_plugin.py [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+        data = toml.load(f)
+    classfier_ls = data["project"]["classifiers"]
+    for classifier in classfier_ls:
+        if classifier.startswith(classifier_key):
+            version = classifier.split(classifier_key.strip() + " ::")[-1]
+            version_ls.append(version.strip())
+    return version_ls
+
+
+def verify_package_version(toml_file, package_name, classifier_key):
```
[the remark this answers is not in the document]
Comment (RubelMozumder) [reply pynxtools/test_suite/reader_plugin.py:192]: To include the semantic verson of the the `pynxtools` in the plugin `classifier of toml` file. Later on the version of the pynxtools will also be available in th PyPI page. This test will check the conflict of the `pynxtools` (or `nomad-lab`) version written in toml file and  python env.

File: .github/workflows/plugin_test.yaml
[the remark this answers is not in the document]
Comment (RubelMozumder) [reply .github/workflows/plugin_test.yaml:1]: I also thought like you. But I can not get the solution of version conflict possibly from several plugins.

File: pynxtools/test_suite/README.md
[the remark this answers is not in the document]
Comment (RubelMozumder) [reply pynxtools/test_suite/README.md:1]: Definitely this would go there. This is only for review purposes.

- [ ] Move the content from Readme.md to docs.

File: pyproject.toml (hunk 3 above)
Comment (RubelMozumder) [reply pyproject.toml:77]: I kept them out only for testing purposes. They will be repositioned again.

File: pynxtools/test_suite/reader_plugin.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (RubelMozumder) [reply pynxtools/test_suite/reader_plugin.py:197]: Otherwise it is really confusing which version of `pynxtools-foo` of compatible with `pynxtools`.

**Comment by RubelMozumder** [comment]
> TODO list from PR #306
> 
> * [ ]  ~Modify function `test_has_correct_read_func(reader)` in `test_reader.py`~
> * [x]  Add a generic test template on the Plugin template.
> * [x]  Needs to track the semantics version of the NeXus definition and Pynxtools.
> * [x]  Check plugin packaging that must include the example files.
> 
> The implementation of the first point putting as is not needed as we are not testing all the readers-plugin in the `test_reader.py`. My plan was to bring the existing test for readers residing in pynxtools together with the test for plugin readers.
> 
> For the second point, we have to add another matrix for `pynxtools` version in the ci/cd workflow. Then PyPI automatically grap the compatible version for pynxtools. Still, I can not try it in `sts` reader. Once this PR is accepted, I will go and test it to include the semantic version. I have created another PR for that: #319

I add test for semantic versions of pynxtools that will help to know which version of `pynxtools-foo` is compatible with `pynxtools`.

File: pynxtools/test_suite/reader_plugin.py
Comment (RubelMozumder) [reply pynxtools/test_suite/reader_plugin.py:1]: Renamed!

**Comment by RubelMozumder** [comment]
@lukaspie, You are right, we must have the same version of pynxtoolos in `classifier` and `consistency_with_pynxtools`. The version from the classifier will also be found in the `PyPI` package site, as currently it mentions different compatible the versions of python.

**Comment by RubelMozumder** [comment]
@lukaspie would you check if this branch is working with XPS? 

@domna and @lukaspie, this is ready for review and merge.

**Comment by RubelMozumder** [comment]
@domna, I already removed the code from test framework and doc for `pynxtools` version test.

File: pynxtools/test_suite/reader_plugin.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (RubelMozumder) [reply pynxtools/test_suite/reader_plugin.py]: The verification function from the test has been removed.

File: pyproject.toml (hunk 3 above)
Comment (RubelMozumder) [reply pyproject.toml:79]: They are back to place.

File: pyproject.toml [hunk 5]
```diff
@@ -71,12 +71,6 @@ apm = [
 em = [
     "pynxtools-em",
 ]
-mpes = [
```
Comment (sanbrock) [inline pyproject.toml:74]: these should not be removed, otherwise the plugins will not be automatically installed (e.g. in NOMAD)

Comment (RubelMozumder) [reply pyproject.toml:74]: It was not intended to remove any of the plugin. Unfortunately it occurred in behind the scene probably while rebasing. I added it back.

**Comment by RubelMozumder** [comment]
@lukaspie, I already release a `pynxtools-stm` version tag. Now, the new tag now compatible with new validation method. You can test it for `xps` and feel free to merge it. Once it is merge I can also use the general test on the main branch of `stm`.

File: docs/how-tos/using-pynxtools-test-framework.md [hunk 6]
```diff
@@ -0,0 +1,56 @@
+# Generalized Test Functionality for `pynxtools` plugin
+The `pynxtools` sub-package `testing` is to be utilized to write automated test for `pynxtools` reader plugins without having any deeper knowledge of the `pynxtools` internal architecture. The tool mainly supports generalised test for all the reader plugins disregarding the technical details of raw file and internal design of the plugin (It is assumed that the plugin has followed the [plugin template](https://github.com/FAIRmat-NFDI/pynxtools-plugin-template)).
```
Comment (lukaspie) [inline docs/how-tos/using-pynxtools-test-framework.md:2]:
```suggestion
# Generalized Test Functionality for `pynxtools` plugins
```suggestion
The `pynxtools` sub-package `testing` is to be utilized to write automated tests for `pynxtools` reader plugins without requiring in-depth knowledge of the `pynxtools` internal architecture. The tool supports generalised a general test for all reader plugins, irrespective of the technical details of the raw data files and the internal design of the plugin (note: it is assumed that the plugin was built from the [plugin template](https://github.com/FAIRmat-NFDI/pynxtools-plugin-template) or has the same structure internally).
```
```

File: docs/how-tos/using-pynxtools-test-framework.md [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+
+## Why it is needed
+To test plugin integration with the `pynxtools` core system we need to
+1. Test the plugin's integration with `pynxtools` from the plugin's ci/cd.
+2. Test from pynxtools as pynxtools's ci/cd if plugin has been integrated with `pynxtools` properly.
+
```
Comment (lukaspie) [inline docs/how-tos/using-pynxtools-test-framework.md:8]:
```suggestion
## Why it is needed
To test integration of a plugin with the `pynxtools` core system, we need to
1. Test the plugin's integration with `pynxtools` from the plugin's CI/CD.
2. Test in the pynxtools's CI/CD if the plugin has been integrated with `pynxtools` properly.
```

File: docs/how-tos/using-pynxtools-test-framework.md [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+## How to write integration-test for the reader plugin with `pynxtools.testing`
```
Comment (lukaspie) [inline docs/how-tos/using-pynxtools-test-framework.md:9]:
```suggestion
## How to write an integration test for a reader plugin with `pynxtools.testing`
```

File: docs/how-tos/using-pynxtools-test-framework.md [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+
+It is very simple to write a test to verify the plugin integration with `pynxtools` from plugin tests directory. The developer can write the test where they want, only they need to use the provided test interface from `pynxtools`. An example test for `pynxtools-FOO` (a demo plugin name) plugin as follows:
```
Comment (lukaspie) [inline docs/how-tos/using-pynxtools-test-framework.md:11]:
```suggestion
It is very simple to write a test to verify the plugin integration with `pynxtools` within the plugin's tests directory. The developer can place the test where they want, but they need to use the provided test interface from `pynxtools`. An example test for `pynxtools-FOO` (a demo plugin) plugin is given below:
```

File: docs/how-tos/using-pynxtools-test-framework.md [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+# e.g. module_dir = /pynxtools-foo/tests
+module_dir = os.path.dirname(os.path.abspath(__file__))
+
+
+@pytest.mark.parametrize(
+    "nxdl,reader,files_or_dir",
+    [
+        ("NXfoo", FOOReader, f"{module_dir}/../examples/test_data_dir_1"),
+        ("NXfoo", FOOReader, f"{module_dir}/../examples/test_data_dir_2"),
+        ("NXfoo", FOOReader, f"{module_dir}/data/test_data_dir_2"),
```
Comment (lukaspie) [inline docs/how-tos/using-pynxtools-test-framework.md:31]:
```suggestion
        ("NXfoo", FOOReader, f"{module_dir}/../test/data/test_data_dir_1"),
        ("NXfoo", FOOReader, f"{module_dir}/../test/data/test_data_dir_2"),
        ("NXfoo", FOOReader, f"{module_dir}/../test/data/test_data_dir_2"),
```
Test data should be placed within the tests/data folder of each plugin.

File: docs/how-tos/using-pynxtools-test-framework.md [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+    ],
+)
+def test_foo_reader(nxdl, reader, files_or_dir, tmp_path, caplog):
+    """Test for the FooReader or foo reader plugin.
+
+    Parameters
+    ----------
+    nxdl : str
+        Name of the NXDL file e.g. NXfoo from NXfoo.nxdl.xml.
```
Comment (lukaspie) [inline docs/how-tos/using-pynxtools-test-framework.md:40]:
```suggestion
        Name of the NXDL application definition that is to be tested by
        this reader plugin (e.g. NXfoo), without the file ending .nxdl.xml.
```

File: docs/how-tos/using-pynxtools-test-framework.md [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+    reader : sts
+        _description_
```
Comment (lukaspie) [inline docs/how-tos/using-pynxtools-test-framework.md:42]:
```suggestion
    reader : foo
        Name of the class of the reader.
```

File: docs/how-tos/using-pynxtools-test-framework.md [hunk 13]
```diff
@@ -0,0 +1,56 @@
...
+    files_or_dir : class
+        Name of the class of the reader.
```
Comment (lukaspie) [inline docs/how-tos/using-pynxtools-test-framework.md:44]:
```suggestion
    files_or_dir : class
        est data directory that contains all the files required for running the test data conversion..
```

File: docs/how-tos/using-pynxtools-test-framework.md [hunk 14]
```diff
@@ -0,0 +1,56 @@
...
+    tmp_path : pytest.fixture
+        Pytest fixture parameters.
```
Comment (lukaspie) [inline docs/how-tos/using-pynxtools-test-framework.md:46]:
```suggestion
    tmp_path : pytest.fixture
        Pytest fixture variable, used to clean up the files generated during
        the test.
```

File: docs/how-tos/using-pynxtools-test-framework.md [hunk 15]
```diff
@@ -0,0 +1,56 @@
...
+    caplog : pytest.fixture
+        Log capture from pytest.
```
Comment (lukaspie) [inline docs/how-tos/using-pynxtools-test-framework.md:48]:
```suggestion
        Pytest fixture variable, used to capture the log messages during the test.
```

File: docs/how-tos/using-pynxtools-test-framework.md [hunk 16]
```diff
@@ -0,0 +1,56 @@
...
+    """
+    # test plugin reader
+    test = ReaderTest(nxdl, reader, files_or_dir, tmp_path, caplog)
+    test.convert_to_nexus()
+    test.check_reproducibility_of_nexus()
+```
+
+Alongside the examples in the `examples`, it is also possible to add other examples inside the test directory of the plugin.
```
Comment (lukaspie) [inline docs/how-tos/using-pynxtools-test-framework.md:56]: Should here be some descriptive text about `ignore_undocumented`? I suggest:

"It is also possible to pass the boolean `ignore_undocumented` to `test.convert_to_nexus`. If true, any undocumented keys are ignored in the verification and it is simply checked if the required fields are properly set."

File: docs/index.md [hunk 17]
```diff
@@ -32,6 +32,7 @@ How-to guides provide step-by-step instructions for a wide range of tasks:
 - [Writing an application definition](how-tos/writing-an-appdef.md)
 - [Storing data in multiple application definitions](how-tos/using-multiple-appdefs.md)
 - [Representing experimental geometries](how-tos/transformations.md)
+- [Using pynxtools test framework](using-pynxtools-test-framework.md)
```
Comment (lukaspie) [inline docs/index.md:35]:
```suggestion
- [Using pynxtools test framework](how-tos/using-pynxtools-test-framework.md)
```

File: pynxtools/testing/README.md [hunk 18]
```diff
@@ -0,0 +1,2 @@
+# Generalized Test Functionality for `pynxtools` plugin
+Please follow the documentation for the `Using Pynxtools Test Framework` in `How-Tos` section.
```
Comment (lukaspie) [inline pynxtools/testing/README.md:2]: You can make this a direct link to the docs:
```suggestion
The sub-package `testing` is to be utilized to write automated test for `pynxtools` reader plugins without having any deeper knowledge of the `pynxtools` internal architecture.
Please follow the documentation on [using the pynxtools test framework](https://fairmat-nfdi.github.io/pynxtools/how-tos/using-pynxtools-test-framework.html).
```

File: pyproject.toml [hunk 19]
```diff
@@ -65,18 +65,15 @@ dev = [
...
 apm = [
     "pynxtools-apm",
 ]
 em = [
     "pynxtools-em",
 ]
-mpes = [
-    "pynxtools-mpes",
-]
-xps = [
```
Comment (lukaspie) [inline pyproject.toml:77]: Please enable the xps plugin again here.

**Review by lukaspie — CHANGES_REQUESTED** [review]
Mostly typos fixes

File: docs/how-tos/using-pynxtools-test-framework.md (hunk 6 above)
Comment (RubelMozumder) [reply docs/how-tos/using-pynxtools-test-framework.md]: Noted!

File: docs/how-tos/using-pynxtools-test-framework.md (hunk 7 above)
Comment (RubelMozumder) [reply docs/how-tos/using-pynxtools-test-framework.md]: Noted!

File: docs/how-tos/using-pynxtools-test-framework.md (hunk 10 above)
Comment (RubelMozumder) [reply docs/how-tos/using-pynxtools-test-framework.md]: Noted!

File: docs/how-tos/using-pynxtools-test-framework.md (hunk 11 above)
Comment (RubelMozumder) [reply docs/how-tos/using-pynxtools-test-framework.md]: Noted!

File: pyproject.toml (hunk 19 above)
Comment (RubelMozumder) [reply pyproject.toml:77]: Noted!

File: docs/how-tos/using-pynxtools-test-framework.md [hunk 20]
```diff
@@ -0,0 +1,56 @@
+# Generalized Test Functionality for `pynxtools` plugins
+The pynxtools sub-package testing is to be utilized to write automated tests for pynxtools reader plugins without requiring in-depth knowledge of the pynxtools internal architecture. The tool supports generalised a general test for all reader plugins, irrespective of the technical details of the raw data files and the internal design of the plugin (note: it is assumed that the plugin was built from the [plugin template](https://github.com/FAIRmat-NFDI/pynxtools-plugin-template) or has the same structure internally).
```
Comment (lukaspie) [inline docs/how-tos/using-pynxtools-test-framework.md:2]: Why did you remove the quotes here? I would go with 
"The `pynxtools` sub-package `testing` is to be utilized to write automated test for `pynxtools`..."

File: docs/how-tos/using-pynxtools-test-framework.md [hunk 21]
```diff
@@ -0,0 +1,56 @@
...
+    """
+    # test plugin reader
+    test = ReaderTest(nxdl, reader, files_or_dir, tmp_path, caplog)
+    test.convert_to_nexus()
+    # Use `ignore_undocumented` to skip undocumented fields
+    # test.convert_to_nexus(ignore_undocumented=True)
+    test.check_reproducibility_of_nexus()
+```
+
+Alongside the examples in the `examples`, it is also possible to add other examples inside the test directory of the plugin. It is also possible to pass the boolean ignore_undocumented to test.convert_to_nexus. If true, any undocumented keys are ignored in the verification and it is simply checked if the required fields are properly set.
```
Comment (lukaspie) [inline docs/how-tos/using-pynxtools-test-framework.md:56]:
```suggestion
Alongside the examples in the `examples`, it is also possible to add other examples inside the test directory of the plugin. It is also possible to pass the boolean `ignore_undocumented` to `test.convert_to_nexus`. If true, any undocumented keys are ignored in the verification and it is simply checked if the required fields are properly set.
```

File: docs/how-tos/using-pynxtools-test-framework.md (hunk 20 above)
Comment (RubelMozumder) [reply docs/how-tos/using-pynxtools-test-framework.md:2]: Noted!

File: docs/how-tos/using-pynxtools-test-framework.md (hunk 21 above)
Comment (RubelMozumder) [reply docs/how-tos/using-pynxtools-test-framework.md:56]: Noted!

File: docs/index.md (hunk 17 above)
Comment (RubelMozumder) [reply docs/index.md:35]: Noted!

File: docs/how-tos/using-pynxtools-test-framework.md (hunk 9 above)
Comment (RubelMozumder) [reply docs/how-tos/using-pynxtools-test-framework.md]: Noted!

File: docs/how-tos/using-pynxtools-test-framework.md (hunk 8 above)
Comment (RubelMozumder) [reply docs/how-tos/using-pynxtools-test-framework.md]: Noted!

File: docs/how-tos/using-pynxtools-test-framework.md (hunk 13 above)
Comment (RubelMozumder) [reply docs/how-tos/using-pynxtools-test-framework.md:41]: Noted!

File: .github/workflows/plugin_test.yaml [hunk 22]
```diff
@@ -0,0 +1,102 @@
+# This workflow will install Python dependencies, run tests and lint with a single version of Python
+# For more information see: https://help.github.com/actions/language-and-framework-guides/using-python-with-github-actions
+
+name: test plugin
+
+on:
+  push:
+    branches: [generic_test_in_ci_cd]
```
Comment (lukaspie) [inline .github/workflows/plugin_test.yaml:8]: This should probably changed to master

Comment (RubelMozumder) [reply .github/workflows/plugin_test.yaml:8]: Thanks! To bring this. I that should be `*` which means all in push. This is for just now. If you approve it I will change it immediately.

File: .github/workflows/plugin_test.yaml [hunk 23]
```diff
@@ -0,0 +1,102 @@
...
+      #   with:
+      #     fetch-depth: 0
+      #     repository: FAIRmat-NFDI/pynxtools-apm
+      #     path: pynxtools-apm
+      # - name: Run apm tests
+      #   run: |
+      #     cd pynxtools-apm
+      #     git checkout main
+      #     pip install .
+      #     pytest tests/.
```
Comment (lukaspie) [inline .github/workflows/plugin_test.yaml:102]: Please at the commented test for pynx-ellips to plugin_test.yaml:
```suggestion
      #     pytest tests/.
      # # Test for ellipsometry Plugin
      # - name: clone ellipsometry repo
      #   uses: actions/checkout@v4
      #   with:
      #     fetch-depth: 0
      #     repository: FAIRmat-NFDI/pynxtools-ellips
      #     path: pynxtools-ellips
      # - name: Run ellipsometry tests
      #   run: |
      #     cd pynxtools-ellips
      #     git checkout main
      #     pip install .
      #     pytest tests/.
```

File: .github/workflows/plugin_test.yaml (hunk 22 above)
Comment (RubelMozumder) [reply .github/workflows/plugin_test.yaml]: NOted!

**Review by lukaspie — APPROVED** [review]
LGTM. Let's first merge this and then we can see what we want to keep from #331
